### PR TITLE
Enable little-endian signing and dual-endian verification in SPDM 1.0…

### PR DESCRIPTION
**Pull request by richkong88**
…\1.1

Fix: https://github.com/DMTF/libspdm/issues/2151

### CODE REVIEW

**Comment by jyao1** [comment]
@richkong88 , would you please squash the commit? There is no need to add patch for coding style fix.

File: include/library/spdm_lib_config.h [hunk 1]
```diff
@@ -383,4 +383,68 @@
 #define LIBSPDM_CHECK_SPDM_CONTEXT 1
 #endif
 
+/* Enable Endianness for signatures on SPDM1.0 and SPDM1.1.
+* Big-endian is default, but little-endian may need to be supported as well.
+* See issue: https://github.com/DMTF/libspdm/issues/2151 */
```
Comment (jyao1) [inline include/library/spdm_lib_config.h:388]: I dont think we need this line.

**Comment by richkong88** [comment]
> @richkong88 , would you please squash the commit? There is no need to add patch for coding style fix.

Yes, I'll put everything into a single CL, once all feedback is incorporated.

**Comment by jyao1** [comment]
I think we need to update req_asym version as well, such as `libspdm_req_asym_sign`, `libspdm_req_asym_verify`.

File: library/spdm_crypt_lib/libspdm_crypt_asym.c [hunk 2]
```diff
@@ -833,28 +940,97 @@ bool libspdm_asym_verify_hash(
...
     }
 }
 
+void libspdm_asym_signature_swap_endian_if_necessary(
+    spdm_version_number_t spdm_version, uint32_t base_asym_algo,
+    uint8_t *signature, size_t sig_size)
+{
+#if LIBSPDM_SPDM_10_11_SIGN_LITTLE_ENDIAN_RSA_ECDSA
+    uint32_t swap_endian_rsa = 0;
+    uint32_t swap_endian_ecdsa = 0;
```
Comment (jyao1) [inline library/spdm_crypt_lib/libspdm_crypt_asym.c:977]: I think swap_endian_rsa and swap_endian_ecdsa should be a `bool` type.

File: library/spdm_crypt_lib/libspdm_crypt_asym.c [hunk 3]
```diff
@@ -589,12 +589,49 @@ bool libspdm_sm2_dsa_verify_wrap (void *context, size_t hash_nid,
 }
 #endif
 
+static void libspdm_copy_signature_swap_endian(
+    uint8_t *endian_swapped_signature_buffer,
+    size_t endian_swapped_signature_buffer_size,
+    bool is_dual_buffer,
```
Comment (jyao1) [inline library/spdm_crypt_lib/libspdm_crypt_asym.c:595]: `is_dual_buffer` brings confusion to me, and it takes me sometime to understand what it really means.


Maybe just define two functions: libspdm_copy_signature_swap_endian_for_ecdsa() and libspdm_copy_signature_swap_endian_for_rsa() ?

Then those functions can be reused by libspdm_asym_signature_swap_endian_if_necessary().

File: library/spdm_crypt_lib/libspdm_crypt_asym.c (hunk 2 above)
Comment (richkong88) [reply library/spdm_crypt_lib/libspdm_crypt_asym.c:977]: Agree.

**Comment by richkong88** [comment]
> I think we need to update req_asym version as well, such as `libspdm_req_asym_sign`, `libspdm_req_asym_verify`.
> 
> Also, can we have some unit test or CI to ensure there will be no regression?

Will update req_asym versions and add unit tests.

File: library/spdm_crypt_lib/libspdm_crypt_asym.c (hunk 3 above)
Comment (richkong88) [reply library/spdm_crypt_lib/libspdm_crypt_asym.c:595]: Agree. I split into rsa and ecdsa versions and tried to make the code clearer.

File: include/library/spdm_lib_config.h (hunk 1 above)
Comment (richkong88) [reply include/library/spdm_lib_config.h:388]: Removed.

**Comment by steven-bellock** [comment]
In general the macros should only be used if they impart a significant reduction in size. In this case they don't so I would like to see this set at runtime through `libspdm_set_data`. Also I don't think we need to differentiate between 1.0 and 1.1. Just lump them together. I can't imagine a Responder that would use LE for 1.0 and BE for 1.1.

**Comment by richkong88** [comment]
> In general the macros should only be used if they impart a significant reduction in size. In this case they don't so I would like to see this set at runtime through `libspdm_set_data`. 

While I'm not generally opposed to a runtime decision, it seems that the libspdm_crypt_asym routines do not receive an "spdm_context" to extract the libspdm_set_data field? I think any "context" parameters are crypto contexts.  So making it runtime would involve changing a lot of interfaces (to take another parameter to indicate the runtime choic) which I was trying to minimize. Please correct me if I'm misunderstanding this.

> Also I don't think we need to differentiate between 1.0 and 1.1. Just lump them together. I can't imagine a Responder that would use LE for 1.0 and BE for 1.1.

I'm fine lumping 1.0 and 1.1 together. @jyao1 Jiewen, do you think lumping them together is ok?

**Comment by jyao1** [comment]
If we use runtime mechanism, we can combine 1.0 and 1.1 because the integrator can make final decision based upon version.

But if we want to keep MACRO, then I think we should separate them to allow the flexibility, because the text in 1.0 and 1.1 are different.

**Comment by richkong88** [comment]
Going to try a runtime mechanism. Closing this pull request.

**Pull request closed without merging**